File: services/scraper/app/parsing.py
from urllib.parse import urlparse

from app.extractors import ExtractionResult, extract_generic_product
from app.extractors.stores import extract_store_product
from app.models import QualityResult
from app.quality import evaluate_quality


def parse_product_html(
    html_text: str,
    page_url: str,
) -> tuple[ExtractionResult, QualityResult]:
    store_extraction = extract_store_product(html_text, page_url)
    generic_extraction = extract_generic_product(html_text, page_url)
    if store_extraction is None:
        extraction = generic_extraction
    else:
        blocked_merge_fields = {
            warning.split(":", 1)[1]
            for warning in store_extraction.warnings
            if warning.startswith("do_not_merge:")
        }
        for field_name in blocked_merge_fields:
            if field_name in generic_extraction.product.model_fields:
                setattr(generic_extraction.product, field_name, None)
                generic_extraction.sources.pop(field_name, None)
        extraction = store_extraction
        extraction.merge_missing(generic_extraction)
    _remove_store_placeholders(extraction, page_url)
    quality = evaluate_quality(extraction)
    return extraction, quality


def _remove_store_placeholders(
    extraction: ExtractionResult,
    page_url: str,
) -> None:
    hostname = (urlparse(page_url).hostname or "").lower()
    title = (extraction.product.title or "").strip().lower()
    blocked_titles: dict[str, set[str]] = {
        "aliexpress.": {"aliexpress", "aliexpress.com"},
        "ebay.": {"error page | ebay"},
        "emag.": {"javascript is disabled"},
        "takealot.com": {"takealot.com: online shopping | sa's leading online store"},
        "cdiscount.com": {"cdiscount.com"},
        "farfetch.com": {
            "register & sign in - farfetch",
            "register &amp; sign in - farfetch",
        },
    }
    matched_domain = next(
        (domain for domain in blocked_titles if domain in hostname),
        None,
    )
    if matched_domain and title in blocked_titles[matched_domain]:
        for field_name in extraction.product.model_fields:
            if field_name == "has_discount":
                setattr(extraction.product, field_name, False)
            else:
                setattr(extraction.product, field_name, None)
            extraction.sources.pop(field_name, None)
        extraction.warnings.append(f"{matched_domain.rstrip('.')}_non_product_page")
        return
    if "aliexpress." in hostname and title in {"aliexpress", "aliexpress.com"}:
        extraction.product.title = None
        extraction.sources.pop("title", None)
        extraction.warnings.append("aliexpress_placeholder_title")
    description = (extraction.product.description or "").strip()
    if "aliexpress." in hostname and description.lower().startswith("smarter shopping"):
        extraction.product.description = None
        extraction.sources.pop("description", None)
        extraction.warnings.append("aliexpress_placeholder_description")

File: packages/scraper-core/src/parsing.ts
/** Port of `services/scraper/app/parsing.py`. */
import { parseHtml } from "./dom";
import { extractGenericProduct } from "./generic";
import { hostnameOf } from "./normalization";
import { evaluateQuality } from "./quality";
import { mergeMissing } from "./result";
import { extractStoreProduct } from "./stores/registry";
import { PRODUCT_FIELDS, type ExtractionResult, type QualityResult } from "./types";

export type ParseOutcome = {
  extraction: ExtractionResult;
  quality: QualityResult;
};

/**
 * Parses an already-fetched page. `document` is either a live `WebView`
 * document or one produced by `DOMParser`; `htmlText` is the matching source
 * used by the regex-only passes.
 */
export function parseProductDocument(
  document: Document,
  htmlText: string,
  pageUrl: string,
): ParseOutcome {
  const storeExtraction = extractStoreProduct(document, htmlText, pageUrl);
  const genericExtraction = extractGenericProduct(document, htmlText, pageUrl);

  let extraction: ExtractionResult;
  if (storeExtraction === null) {
    extraction = genericExtraction;
  } else {
    const blockedMergeFields = new Set(
      storeExtraction.warnings
        .filter((warning) => warning.startsWith("do_not_merge:"))
        .map((warning) => warning.slice("do_not_merge:".length)),
    );
    for (const field of blockedMergeFields) {
      if ((PRODUCT_FIELDS as readonly string[]).includes(field)) {
        (genericExtraction.product as Record<string, unknown>)[field] = null;
        delete genericExtraction.sources[field];
      }
    }
    extraction = storeExtraction;
    mergeMissing(extraction, genericExtraction);
  }

  removeStorePlaceholders(extraction, pageUrl);
  return { extraction, quality: evaluateQuality(extraction) };
}

/** Convenience wrapper that parses an HTML string (used by the fetch tier). */
export function parseProductHtml(htmlText: string, pageUrl: string): ParseOutcome {
  return parseProductDocument(parseHtml(htmlText), htmlText, pageUrl);
}

const PLACEHOLDER_TITLES: Record<string, string[]> = {
  "aliexpress.": ["aliexpress", "aliexpress.com"],
  "ebay.": ["error page | ebay"],
  "emag.": ["javascript is disabled"],
  "takealot.com": ["takealot.com: online shopping | sa's leading online store"],
  "cdiscount.com": ["cdiscount.com"],
  "farfetch.com": ["register & sign in - farfetch", "register &amp; sign in - farfetch"],
};

function removeStorePlaceholders(extraction: ExtractionResult, pageUrl: string): void {
  const hostname = hostnameOf(pageUrl);
  const title = (extraction.product.title ?? "").trim().toLowerCase();
  const matchedDomain = Object.keys(PLACEHOLDER_TITLES).find((domain) => hostname.includes(domain));

  if (matchedDomain && PLACEHOLDER_TITLES[matchedDomain].includes(title)) {
    for (const field of PRODUCT_FIELDS) {
      (extraction.product as Record<string, unknown>)[field] =
        field === "has_discount" ? false : null;
      delete extraction.sources[field];
    }
    extraction.warnings.push(`${matchedDomain.replace(/\.+$/, "")}_non_product_page`);
    return;
  }

  const description = (extraction.product.description ?? "").trim();
  if (
    hostname.includes("aliexpress.") &&
    description.toLowerCase().startsWith("smarter shopping")
  ) {
    extraction.product.description = null;
    delete extraction.sources.description;
    extraction.warnings.push("aliexpress_placeholder_description");
  }
}
